dataset/jeopardy_dataset.py

```diff
@@ -0,0 +1,203 @@
+import json
+from PIL import Image
+from torch.utils.data import Dataset
+import torch
+import os
+import string
+from tqdm import tqdm
+from transformers import RobertaTokenizer
+from collections import Counter
+from transformers.data.data_collator import DataCollatorForLanguageModeling
+
+UNK_TOKEN = "<unk>"
+
+# TODO: no restriction on answers
+class JeopardyDataset(Dataset):
+    def __init__(self, 
+        questions_file=None, 
+        answers_file=None, 
+        images_dir=None, 
+        transform=None, 
+        train=True, 
+        q_len=8, 
+        ans_len=1, 
+        test_split=0.2,
+        frequency_threshold=8, 
+        multiple_images=False):
+        """
+        Args:
+            questions_file (string): Path to the json file with questions.
+            answers_file (string): Path to the json file with annotations.
+            images_dir (string): Directory with all the images.
+            transform (callable, optional): Optional transform to be applied
+                on a sample.
+            frequency_threshold: how many times does a word have to appear as an answer for it to be considered. (default: 8) 
+            As used in Teney et al. https://arxiv.org/pdf/1708.02711.pdf
+                
+        Example call:
+            test_ds = JeopardyDataset("v2_OpenEnded_mscoco_train2014_questions.json", "v2_mscoco_train2014_annotations.json", "images")
+
+        """
+        self.multiple_images = multiple_images
+        # initializing the lengths of our questions and answers
+        self.q_len = q_len
+        self.ans_len = ans_len
+        # return test or train set?
+        self.train = train
+
+        q_f = open(questions_file, 'r')
+        a_f = open(answers_file, 'r')
+        
+        # we want only 80% of the question/answer text to build the vocabulary
+        self.answers = json.load(a_f)["annotations"]
+        self.questions = json.load(q_f)
+
+        # don't need to worry about shuffling here because the questions
+        # in the file are in random order (ctsy utils.model_utils:shuffled_questions_file()) 
+        split_index = int((1-test_split) * len(self.questions))
+        
+        if train:
+            self.questions = self.questions[:split_index]
+            self.answers = self.answers[:split_index]
+        else:
+            self.questions = self.questions[split_index:]
+            self.answers = self.answers[split_index:]
+            
+        questions_dict = {q['question_id']: (q["question"], int(q["image_id"])) for q in self.questions}
+        self.images_dir = images_dir
+
+        self.image_id_to_filename = self._find_images()
+        self.transform = transform    
+        self.frequent_answers = self._find_frequent_answers(frequency_threshold)
+        self.question_to_answer = self._find_answers()
+        self.tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
+
+        # filter down questions_dict based on the answer set self.frequent_answers
```

Comment:
is this true? `frequent_answers` does not appear in the `filter_questions_dict` object